src/main/java/net/consensys/orion/http/handler/privacy/AddToPrivacyGroupHandler.java

```diff
@@ -0,0 +1,152 @@
+                oldPrivacyGroupPayload.randomSeed()));
+        Stream<Box.PublicKey> combinedAddresses = Arrays
+            .stream(combinedPrivacyGroup.get().addresses())
+            .filter(key -> !key.equals(modifyPrivacyGroupRequest.from()))
+            .distinct()
+            .map(enclave::readKey);
+
+        final var addRequests = sendRequestsToOthers(
+            combinedAddresses,
+            new SetPrivacyGroupRequest(combinedPrivacyGroup.get(), modifyPrivacyGroupRequest.privacyGroupId()),
+            "/setPrivacyGroup");
+
+        CompletableFuture.allOf(addRequests).whenComplete((iAll, iEx) -> {
+          if (iEx != null) {
+            handleFailure(routingContext, iEx);
+            return;
+          }
+
+          /*
+           * Todo here: get ordered list of commitments and private transactions from `PrivateTransactionStore` to push to Pantheon
+           * */
+
+          final PrivacyGroupPayload innerCombinedPrivacyGroupPayload = combinedPrivacyGroup.get();
+          privacyGroupStorage
+              .update(privacyGroupStorage.generateDigest(oldPrivacyGroupPayload), innerCombinedPrivacyGroupPayload)
+              .thenAccept((privacyGroupResult) -> {
+
+                final QueryPrivacyGroupPayload queryPrivacyGroupPayload =
+                    new QueryPrivacyGroupPayload(innerCombinedPrivacyGroupPayload.addresses(), null);
+
+                queryPrivacyGroupPayload.setPrivacyGroupToAppend(modifyPrivacyGroupRequest.privacyGroupId());
+                final String key = queryPrivacyGroupStorage.generateDigest(queryPrivacyGroupPayload);
+
+                log.info("Stored privacy group. resulting digest: {}", key);
+                queryPrivacyGroupStorage
+                    .update(key, queryPrivacyGroupPayload)
+                    .thenAccept((queryPrivacyGroupStorageResult) -> {
+                      final PrivacyGroup group = new PrivacyGroup(
+                          modifyPrivacyGroupRequest.privacyGroupId(),
+                          PrivacyGroupPayload.Type.PANTHEON,
+                          innerCombinedPrivacyGroupPayload.name(),
+                          innerCombinedPrivacyGroupPayload.description(),
+                          innerCombinedPrivacyGroupPayload.addresses());
+                      log.info("Storing privacy group {} complete", modifyPrivacyGroupRequest.privacyGroupId());
+
+                      final Buffer toReturn = Buffer.buffer(Serializer.serialize(JSON, group));
+                      routingContext.response().end(toReturn);
+                    })
+                    .exceptionally(
+                        e -> routingContext
+                            .fail(new OrionException(OrionErrorCode.ENCLAVE_UNABLE_STORE_PRIVACY_GROUP, e)));
+              })
+              .exceptionally(
+                  e -> routingContext.fail(new OrionException(OrionErrorCode.ENCLAVE_UNABLE_STORE_PRIVACY_GROUP, e)));
+        });
+      }
+    }).exceptionally(
```

Comment:
Looks like if this handler is invoked "illegally" (eg a non-member attempting to add someone) it results in a http response code of 500 (internal_server_error) - is that expected? I'd have thought maybe a 403 (forbidden)?

Reply:
As a result of this comment I realized that the test for this didn't work properly at all, so I've fixed that. As to the 403 vs 500 - returning not found when someone unauthorized tries to add themselves doesn't acknowledge that the group with the given ID even exists, which leads to less information disclosure. I think there are merits to both approaches.